### PR TITLE
[New Scheduler] Add DataManagementService

### PR DESCRIPTION
**Pull request by style95**
## Description
This component is in charge of storing data to ETCD.
It is based on eventual consistency.
If it fails to store data for some reason, it keeps retrying until data is stored.

## Related issue and scope

- [ ] I opened an issue to propose and discuss this change (#????)

## My changes affect the following components

- [ ] API
- [ ] Controller
- [ ] Message Bus (e.g., Kafka)
- [ ] Loadbalancer
- [ ] Invoker
- [ ] Intrinsic actions (e.g., sequences, conductors)
- [ ] Data stores (e.g., CouchDB)
- [ ] Tests
- [ ] Deployment
- [ ] CLI
- [ ] General tooling
- [ ] Documentation

## Types of changes

- [ ] Bug fix (generally a non-breaking change which closes an issue).
- [x] Enhancement or new feature (adds new functionality).
- [ ] Breaking change (a bug fix or enhancement which changes existing behavior).

## Checklist:

- [x] I signed an [Apache CLA](https://github.com/apache/openwhisk/blob/master/CONTRIBUTING.md).
- [x] I reviewed the [style guides](https://github.com/apache/openwhisk/wiki/Contributing:-Git-guidelines#code-readiness) and followed the recommendations (Travis CI will check :).
- [x] I added tests to cover my changes.
- [ ] My changes require further changes to the documentation.
- [ ] I updated the documentation where necessary.

### CODE REVIEW

File: common/scala/src/main/scala/org/apache/openwhisk/core/service/LeaseKeepAliveService.scala
Comment (style95) [inline common/scala/src/main/scala/org/apache/openwhisk/core/service/LeaseKeepAliveService.scala:14]: I realized this pr is also dependent on these modules.
It would be better to merge this PR after this kind of module is introduced.

File: common/scala/src/main/scala/org/apache/openwhisk/core/service/DataManagementService.scala
Comment (bdoyle0182) [inline common/scala/src/main/scala/org/apache/openwhisk/core/service/DataManagementService.scala:159]: Should change to configuration before merging. Is 1 second a good default for this?

File: common/scala/src/main/scala/org/apache/openwhisk/core/service/DataManagementService.scala
Comment (bdoyle0182) [inline common/scala/src/main/scala/org/apache/openwhisk/core/service/DataManagementService.scala:143]: nit: don't think this comment really adds anything

File: common/scala/src/main/scala/org/apache/openwhisk/core/service/DataManagementService.scala
Comment (bdoyle0182) [inline common/scala/src/main/scala/org/apache/openwhisk/core/service/DataManagementService.scala:135]:
```suggestion
    // It should not receive "prefixed" data
```

File: common/scala/src/main/scala/org/apache/openwhisk/core/service/WatcherService.scala
Comment (bdoyle0182) [inline common/scala/src/main/scala/org/apache/openwhisk/core/service/WatcherService.scala:9]:
```suggestion
case class UnwatchEndpoint(watchKey: String, isPrefix: Boolean, watchName: String, needFeedback: Boolean = false)
```

File: common/scala/src/main/scala/org/apache/openwhisk/core/service/DataManagementService.scala
Comment (bdoyle0182) [inline common/scala/src/main/scala/org/apache/openwhisk/core/service/DataManagementService.scala:248]: what data is stored?

File: common/scala/src/main/scala/org/apache/openwhisk/core/service/DataManagementService.scala
Comment (bdoyle0182) [inline common/scala/src/main/scala/org/apache/openwhisk/core/service/DataManagementService.scala:299]: akka should have a built in `scheduleOnce` method rather than have your own method right?

i.e. `actorSystem.scheduler.scheduleOnce(50 milliseconds, this, "foo")`

File: common/scala/src/main/scala/org/apache/openwhisk/core/service/LeaseKeepAliveService.scala
Comment (bdoyle0182) [inline common/scala/src/main/scala/org/apache/openwhisk/core/service/LeaseKeepAliveService.scala:9]:
```suggestion
case object RegrantLease
```

File: common/scala/src/main/scala/org/apache/openwhisk/core/service/DataManagementService.scala
Comment (bdoyle0182) [inline common/scala/src/main/scala/org/apache/openwhisk/core/service/DataManagementService.scala:172]: Is the parent always `dataManagementService`? If so it would help to be more descriptive than `parent`

Comment (bdoyle0182) [inline common/scala/src/main/scala/org/apache/openwhisk/core/service/DataManagementService.scala:172]: Is the parent always `dataManagementService`? If so it would help to be more descriptive than `parent`

File: common/scala/src/main/scala/org/apache/openwhisk/core/service/DataManagementService.scala
Comment (style95) [reply common/scala/src/main/scala/org/apache/openwhisk/core/service/DataManagementService.scala]: In this case, it stores the request into a buffer because there is already precedent request processing. If any issue happens it would let us know if the request has processed or not.

**Comment by style95** [comment]
I think we need to merge a PR for CI tests for scheduler components before merging this PR.
Since some downstream is running their CI tests based on the upstream core repo, we need to differentiate scheduler tests from the others. Need to see if we can setup another dedicated pipeline for scheduler components in Travis.
And when all components contribution is over, we can merge the pipeline into existing unit/system pipelines.

File: common/scala/src/main/scala/org/apache/openwhisk/core/service/DataManagementService.scala
Comment (style95) [inline common/scala/src/main/scala/org/apache/openwhisk/core/service/DataManagementService.scala:42]: I will add test cases after setting up the CI pipeline for scheduler components.

File: common/scala/src/main/scala/org/apache/openwhisk/core/service/DataManagementService.scala
Comment (bdoyle0182) [reply common/scala/src/main/scala/org/apache/openwhisk/core/service/DataManagementService.scala]: So I'm still learning what this is doing, but what does each request involve? Is it every activation or some sort of metadata setup? If it's every activation it would seem spammy to me otherwise I think it's fine

File: common/scala/src/main/scala/org/apache/openwhisk/core/service/DataManagementService.scala
Comment (style95) [inline common/scala/src/main/scala/org/apache/openwhisk/core/service/DataManagementService.scala:47]: This class is used by both schedulers and invokers to store data to ETCD.
The following kinds of data are stored to ETCD.
1. Throttling data(Action / Namespace)
2. Queue endpoint(where a queue is running)
3. Scheduler endpoint.
4. Container data(running container, warmed container, data to describe how many containers are being created)

Dependent modules are Queue, ContainerProxy, CreationJobManager, etc.

**Comment by style95** [comment]
Just comes up in my mind is it would be great to write down some documents for each component in Wiki.

**Comment by style95** [comment]
Waiting for this PR(https://github.com/apache/openwhisk/pull/5067) to be merged.

**Comment by style95** [comment]
I wrote a document about this module: https://cwiki.apache.org/confluence/display/OPENWHISK/DataManagementService

**Comment by style95** [comment]
It's ready to merge.

**Review by ningyougang — APPROVED** [review]

**Comment by style95** [comment]
@bdoyle0182 Do you have any other comments on this PR?

**Comment by bdoyle0182** [comment]
Just one comment. LGTM

**Pull request merged**
